File: training-data/php/jump_game.php
<?php

function canJump(array $a): bool
{
    $reach = 0;
    foreach ($a as $i => $n) {
        if ($i > $reach) {
            return false;
        }
        $reach = max($reach, $i + $n);
    }
    return true;
}

function minJumps(array $a): ?int
{
    [$jumps, $end, $far] = [0, 0, 0];
    for ($i = 0; $i < count($a) - 1; $i++) {
        $far = max($far, $i + $a[$i]);
        if ($i === $end) {
            if ($far <= $i) {
                return null;
            }
            $jumps++;
            $end = $far;
        }
    }
    return $jumps;
}

echo json_encode([canJump([2, 3, 1, 1, 4]), canJump([3, 2, 1, 0, 4]), minJumps([2, 3, 1, 1, 4]), minJumps([3, 2, 1, 0, 4])]), PHP_EOL;

File: training-data/c/jump_game.c
#include <stdbool.h>
#include <stdio.h>

bool can_jump(const int *a, int n) {
    int reach = 0;
    for (int i = 0; i < n && i <= reach; i++)
        if (i + a[i] > reach) reach = i + a[i];
    return reach >= n - 1;
}

int min_jumps(const int *a, int n) {
    int jumps = 0, end = 0, far = 0;
    for (int i = 0; i < n - 1; i++) {
        if (i + a[i] > far) far = i + a[i];
        if (i == end) {
            if (far <= i) return -1;
            jumps++;
            end = far;
        }
    }
    return jumps;
}

int main(void) {
    int a[] = {2, 3, 1, 1, 4}, b[] = {3, 2, 1, 0, 4};
    printf("%d %d | %d %d\n", can_jump(a, 5), can_jump(b, 5), min_jumps(a, 5), min_jumps(b, 5));
    return 0;
}
